File: mobile/src/api/goals.ts
import { apiFetch } from "./client";

export type GoalType = "reflection" | "mood_tracking" | "self_care" | "custom";
export type GoalPeriod = "daily" | "weekly";

export type Goal = {
  id: number;
  title: string;
  description: string | null;
  goal_type: GoalType;
  target_count: number;
  period: GoalPeriod;
  template_key: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
};

export type GoalTemplate = {
  key: string;
  title: string;
  description: string;
  goal_type: GoalType;
  target_count: number;
  period: GoalPeriod;
};

export type GoalProgress = {
  goal: Goal;
  current_count: number;
  target_count: number;
  progress_percentage: number;
  is_completed: boolean;
  period_start: string;
  period_end: string;
  message: string;
};

export type WeeklyGoalSummary = {
  period_start: string;
  period_end: string;
  active_goals: number;
  completed_goals: number;
  partially_completed_goals: number;
  missed_goals: number;
  overall_completion_percentage: number;
  supportive_message: string;
  goals: GoalProgress[];
};

export type GoalCreate = {
  title: string;
  description?: string | null;
  goal_type: GoalType;
  target_count: number;
  period: GoalPeriod;
  template_key?: string | null;
};

export type GoalUpdate = Partial<
  Pick<Goal, "title" | "description" | "target_count" | "period" | "is_active">
>;

export function getGoalTemplates() {
  return apiFetch<GoalTemplate[]>("/goals/templates", { auth: true });
}

export function getGoals(includeInactive = false) {
  return apiFetch<Goal[]>(
    `/goals?include_inactive=${includeInactive ? "true" : "false"}`,
    { auth: true },
  );
}

export function createGoal(payload: GoalCreate) {
  return apiFetch<Goal>("/goals", {
    method: "POST",
    auth: true,
    body: JSON.stringify(payload),
  });
}

export function updateGoal(goalId: number, payload: GoalUpdate) {
  return apiFetch<Goal>(`/goals/${goalId}`, {
    method: "PATCH",
    auth: true,
    body: JSON.stringify(payload),
  });
}

export function deleteGoal(goalId: number) {
  return apiFetch<void>(`/goals/${goalId}`, { method: "DELETE", auth: true });
}

export function getGoalProgress() {
  return apiFetch<GoalProgress[]>("/goals/progress", { auth: true });
}

export function getWeeklyGoalSummary() {
  return apiFetch<WeeklyGoalSummary>("/goals/weekly-summary", { auth: true });
}

export function completeGoal(goalId: number) {
  return apiFetch<GoalProgress>(`/goals/${goalId}/complete`, {
    method: "POST",
    auth: true,
    body: JSON.stringify({}),
  });
}

File: backend/app/schemas/goals.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

GoalType = Literal["reflection", "mood_tracking", "self_care", "custom"]
GoalPeriod = Literal["daily", "weekly"]


class GoalCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=500)
    goal_type: GoalType
    target_count: int = Field(default=1, ge=1, le=100)
    period: GoalPeriod = "weekly"
    template_key: str | None = Field(default=None, max_length=60)


class GoalUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=500)
    target_count: int | None = Field(default=None, ge=1, le=100)
    period: GoalPeriod | None = None
    is_active: bool | None = None


class GoalResponse(BaseModel):
    id: int
    title: str
    description: str | None = None
    goal_type: GoalType
    target_count: int
    period: GoalPeriod
    template_key: str | None = None
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class GoalCompletionCreate(BaseModel):
    note: str | None = Field(default=None, max_length=500)


class GoalTemplate(BaseModel):
    key: str
    title: str
    description: str
    goal_type: GoalType = "self_care"
    target_count: int = 1
    period: GoalPeriod = "weekly"


class GoalProgressResponse(BaseModel):
    goal: GoalResponse
    current_count: int
    target_count: int
    progress_percentage: float
    is_completed: bool
    period_start: str
    period_end: str
    message: str


class WeeklyGoalSummaryResponse(BaseModel):
    period_start: str
    period_end: str
    active_goals: int
    completed_goals: int
    partially_completed_goals: int
    missed_goals: int
    overall_completion_percentage: float
    supportive_message: str
    goals: list[GoalProgressResponse]
